### PR TITLE
Fix #287955: In Windows Explorer, files with special characters in filepath don't open when double-clicked

### PR DESCRIPTION
**Pull request by Spire42**
Resolves: [#287955](https://musescore.org/en/node/287955)

Fixed a problem that caused MuseScore to fail to open files that contained non-ASCII Unicode characters under Windows.

*Use "x" letter to fill the checkboxes below like [x]*

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

File: main/main.cpp [hunk 1]
```diff
@@ -44,9 +50,43 @@ static void initResources()
...
       {
+      // Force the 8-bit text encoding to UTF-8. This is needed on Windows, which would otherwise default to the local ANSI code page.
+      QTextCodec::setCodecForLocale(QTextCodec::codecForName("UTF-8"));
+
       initResources();
 
       ModulesSetup::instance()->setup();
 
-      return Ms::runApplication(argc, argv);
+#ifdef Q_OS_WIN
```
Comment (Jojo-Schmitz) [inline main/main.cpp]: I believe this is only needed for MSVC, not for MinGW, but I don't know whether it harms there.

Comment (Jojo-Schmitz) [reply main/main.cpp]: `#if (defined (_MSCVER) || defined (_MSC_VER))`

Comment (Spire42) [reply main/main.cpp]: There are no Qt-specific platform checks available specifically for MinGW, so I made sure that the changes would work under both MSVC and MinGW. If you are able to test this with MinGW to confirm, please do.

Comment (Jojo-Schmitz) [reply main/main.cpp]: It does for MSVC. And for Windows + GCC == MinGW

Comment (Spire42) [reply main/main.cpp]: > `#if (defined (_MSCVER) || defined (_MSC_VER))`

Would you like me to change it to that?

Comment (Jojo-Schmitz) [reply main/main.cpp]: Unless we can prove that this change works with MinGW too, yes, please

Comment (Spire42) [reply main/main.cpp]: Done.

**Comment by Jojo-Schmitz** [comment]
Building it now, using both, MinGW and MSVC

File: main/main.cpp [hunk 2]
```diff
@@ -44,9 +50,43 @@ static void initResources()
 
 int main(int argc, char** argv)
       {
+      // Force the 8-bit text encoding to UTF-8. This is needed on Windows, which would otherwise default to the local ANSI code page.
+      QTextCodec::setCodecForLocale(QTextCodec::codecForName("UTF-8"));
```
Comment (Jojo-Schmitz) [inline main/main.cpp]: this too may need to get done conditionally?

**Comment by Jojo-Schmitz** [comment]
Works fine with MSVS and MinGW!
Now I'll try the '#ifdef Q_OS_WIN`...

File: main/main.cpp (hunk 2 above)
Comment (Spire42) [reply main/main.cpp]: I considered it, but I decided that it's a good thing to unconditionally force the encoding to UTF-8 on all platforms, since that's what we always want.

Comment (Jojo-Schmitz) [reply main/main.cpp]: But get anywhere else (anywhere but in MSVC) by default, right?

Comment (Jojo-Schmitz) [reply main/main.cpp]: So the comment at least should say that

Comment (Spire42) [reply main/main.cpp]: OK, will do.

**Comment by Jojo-Schmitz** [comment]
Nope, doesn't even build with MinGW when using '#ifdef Q_OS_WIN`, doesn't seem to like the `std::for_each()`, esp. the use auf `auto` in lambda parameter, due to not supporting C++14

**Comment by Jojo-Schmitz** [comment]
Travis doesn't seem to like it, but for entirelry unrelated reasons

File: main/main.cpp (hunk 2 above)
Comment (Spire42) [reply main/main.cpp]: Done.

File: main/main.cpp [hunk 3]
```diff
@@ -44,9 +50,44 @@ static void initResources()
...
+      // would otherwise default to the local ANSI code page and cause corruption of any non-ANSI Unicode characters in command-line arguments.
+      QTextCodec::setCodecForLocale(QTextCodec::codecForName("UTF-8"));
+
       initResources();
 
       ModulesSetup::instance()->setup();
 
-      return Ms::runApplication(argc, argv);
+#if (defined (_MSCVER) || defined (_MSC_VER))
+      // On Windows, we need to manually retrieve the command-line arguments and convert them from UTF-16 to UTF-8.
```
Comment (Jojo-Schmitz) [inline main/main.cpp]: Here again, MSCV not all of Windows ;-)

Comment (Spire42) [reply main/main.cpp]: Fixed.

**Comment by Jojo-Schmitz** [comment]
Now just squash into just one comment and we're ready to merge IMHO

**Comment by Spire42** [comment]
> Now just squash into just one comment and we're ready to merge IMHO

Done.

**Comment by Jojo-Schmitz** [comment]
I'm (very) happy with it, @dmitrio95 ?

**Comment by Jojo-Schmitz** [comment]
Oh, one thing still: the commit title! Make it same as the PR title

File: main/main.cpp [hunk 4]
```diff
@@ -44,9 +50,44 @@ static void initResources()
...
+            argvUTF8.push_back(arg.data());
+
+      // Don't use the arguments passed to main(), because they're in the local ANSI code page.
+      Q_UNUSED(argc);
+      Q_UNUSED(argv);
+
+      int argcFinal { argcUTF16 };
+      char** argvFinal { argvUTF8.data() };
+#else
+      int argcFinal { argc };
```
Comment (dmitrio95) [inline main/main.cpp]: It would be good to test whether this may lead to crashes at application exit. `Ms::runApplication` takes a *reference* to `argc` rather than a value since it has to pass it to [`QApplication`](https://doc.qt.io/qt-5/qapplication.html#QApplication) constructor. `QApplication` requires that those `argc` and `argv` remain valid for the entire `QApplication` lifecycle so I wonder whether declaring them in `main()` function scope is sufficient for that. Should likely be OK but still better to test.

Comment (Jojo-Schmitz) [reply main/main.cpp]: No crash on exit with MSVC

Comment (Spire42) [reply main/main.cpp]: I just spent time time checking this, and :

`QCoreApplication::exit()` is called inside `QCoreApplication::quit()`, which is called inside `MuseScore::closeEvent()`, which is called several levels inside `QCoreApplication::exec()`, which is called inside `Ms::runApplication()`, which is called in `main()`, where all of `argvUTF8Q`, `argvUTF8`, `argcFinal`, and `argvFinal` are in scope and alive.

Further, the documentation for `QCoreApplication::exec()` states:

“We recommend that you connect clean-up code to the `QCoreApplication::aboutToQuit()` signal, instead of putting it in your application's `main()` function because on some platforms the `exec()` call may not return. For example, on Windows when the user logs off, the system terminates the process after Qt closes all top-level windows. Hence, there is no guarantee that the application will have time to exit its event loop and execute code at the end of the `main()` function after the `exec()` call.”

In other words, I'm pretty confident that this is safe.

Comment (Jojo-Schmitz) [reply main/main.cpp]: Please still fix the commit title

**Comment by Spire42** [comment]
> Oh, one thing still: the commit title! Make it same as the PR title

Done.

**Comment by jthistle** [comment]
I can't remember, but it part of the style guidelines to use `int a = 1;` rather than `int a { 1 };`? Either way that's what way do in the rest of the codebase. @Jojo-Schmitz?

**Comment by Marr11317** [comment]
Both seem quite common in the code base

**Comment by Jojo-Schmitz** [comment]
> > Oh, one thing still: the commit title! Make it same as the PR title
> 
> Done.

No, I don't see it

**Comment by Jojo-Schmitz** [comment]
> I can't remember, but it part of the style guidelines to use `int a = 1;` rather than `int a { 1 };`? Either way that's what way do in the rest of the codebase. @Jojo-Schmitz?

Hmm, I'd prefer the former

**Comment by Spire42** [comment]
> No, I don't see it. Commit **title** still reads "Update main.cpp"

Done for real this time.

> > I can't remember, but it part of the style guidelines to use `int a = 1;` rather than `int a { 1 };`?
> 
> Hmm, I'd prefer the former

The style guidelines don't specify. If both ways are common in the code base, I prefer to use modern uniform initialization syntax and leave it the way it is.

File: main/main.cpp [hunk 5]
```diff
@@ -44,9 +50,44 @@ static void initResources()
...
+
       initResources();
 
       ModulesSetup::instance()->setup();
 
-      return Ms::runApplication(argc, argv);
+#if (defined (_MSCVER) || defined (_MSC_VER))
+      // On MSVC under Windows, we need to manually retrieve the command-line arguments and convert them from UTF-16 to UTF-8.
+      // This prevents data loss if there are any characters that wouldn't fit in the local ANSI code page.
+      int argcUTF16 { 0 };
```
Comment (Howard-C) [inline main/main.cpp]: Please use `=` for initializations.

Comment (Howard-C) [reply main/main.cpp]: Oops, didn't see there're already discussions

Comment (Howard-C) [reply main/main.cpp]: But I guess `=` initialization is still way more common than braces in MuseScore code, so I think it's best to use that.

**Comment by jthistle** [comment]
> The style guidelines don't specify. If both ways are common in the code base, I prefer to use modern uniform initialization syntax and leave it the way it is.

The thing is, they're not both common. Well, weirdly, uniform initialization is common in header files, but barely evident at all in cpp files. I don't know. Personally I'd rather see the former, but I don't thi k

File: main/main.cpp (hunk 5 above)
Comment (Jojo-Schmitz) [reply main/main.cpp]: And compaitible with C and C++ programmers ;-)

**Comment by Spire42** [comment]
OK, you guys win. I converted all instances of uniform initialization to copy initialization.

If this is going to be enforced, it should probably go in the style guide.
